### PR TITLE
JFG-987 Add build in metrics SR and virtual users to main parameters

### PR DESCRIPTION
**Pull request by SokolAndrey**

### CODE REVIEW

File: dbapi/src/main/java/com/griddynamics/jagger/dbapi/dto/MetricNameDto.java [hunk 1]
```diff
@@ -93,6 +93,7 @@ public int hashCode() {
         DURATION,
         STANDARD_METRICS,             /* success rate, iterations, etc */
         VALIDATOR,
-        SESSION_SCOPE_TG             /* custom test-group metrics for session scope */
+        SESSION_SCOPE_TG,             /* custom test-group metrics for session scope */
+        VIRTUAL_USERS
```
Comment (dlatnikov) [inline dbapi/src/main/java/com/griddynamics/jagger/dbapi/dto/MetricNameDto.java]: WTF?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/reporting/SummaryReporter.java [hunk 2]
```diff
@@ -280,9 +279,6 @@ private void fillTestSummaryAndScalabilityData(String sessionId) {
                     workloadTaskData.setThroughput(new BigDecimal(metricsForThisTest.get(metricEntity).getValue()));
                     dataForScalabilityPlots.get(testEntity).put(StandardMetricsNamesUtil.THROUGHPUT_ID, metricsForThisTest.get(metricEntity).getValue());
                 }
-                if (StandardMetricsNamesUtil.getAllVariantsOfMetricName(StandardMetricsNamesUtil.FAIL_COUNT_ID).contains(metricEntity.getMetricId())) {
-                    workloadTaskData.setFailuresCount(metricsForThisTest.get(metricEntity).getValue().intValue());
-                }
                 if (StandardMetricsNamesUtil.getAllVariantsOfMetricName(StandardMetricsNamesUtil.SUCCESS_RATE_ID).contains(metricEntity.getMetricId())) {
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/reporting/SummaryReporter.java]: why you decided to leave success rate here?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/sessioncomparation/workload/WorkloadComparisonResult.java [hunk 3]
```diff
@@ -191,12 +191,6 @@ private WorkloadTaskData createDummyWorkloadTaskData(SessionEntity sessionEntity
         } else {
             workloadTaskData.setThroughput(new BigDecimal(0));
         }
-        Double failuresCount = getStandardMetricValueById(StandardMetricsNamesUtil.FAIL_COUNT_ID,standardMetricsMap).getValue();
-        if (failuresCount != null) {
-            workloadTaskData.setFailuresCount(failuresCount.intValue());
-        } else {
-            workloadTaskData.setFailuresCount(0);
-        }
         Double successRate = getStandardMetricValueById(StandardMetricsNamesUtil.SUCCESS_RATE_ID,standardMetricsMap).getValue();
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/sessioncomparation/workload/WorkloadComparisonResult.java]: why SR is left here?

File: chassis/util/src/main/java/com/griddynamics/jagger/util/StandardMetricsNamesUtil.java [hunk 4]
```diff
@@ -37,23 +45,22 @@
...
     public static final String ITERATION_SAMPLES_ID = "samples";
+    public static final String VIRTUAL_USERS_ID = "Jagger.Threads";
     //end: following section is used for docu generation - standard metrics ids
 
     // ids for standard metrics saved with old model (in WorkloadTaskData, TimeLatencyPercentile, etc)
     public static final String THROUGHPUT_OLD_ID = "throughput-old";
     public static final String LATENCY_OLD_ID = "avgLatency-old";
     public static final String LATENCY_STD_DEV_OLD_ID = "stdDevLatency-old";
-    public static final String FAIL_COUNT_OLD_ID = "failureCount-old";
     public static final String SUCCESS_RATE_OLD_ID = "successRate-old";
```
Comment (dlatnikov) [inline chassis/util/src/main/java/com/griddynamics/jagger/util/StandardMetricsNamesUtil.java]: why is it left? here and everywhere

File: dbapi/src/main/java/com/griddynamics/jagger/dbapi/model/rules/TreeViewGroupMetricsToNodeRuleProvider.java [hunk 5]
```diff
@@ -148,6 +136,15 @@ public TreeViewGroupMetricsToNodeRule provide(Map<String, Set<String>> agentName
                 StandardMetricsNamesUtil.SUCCESS_RATE_ID,
                 StandardMetricsNamesUtil.SUCCESS_RATE,
                 regex));
```
Comment (dlatnikov) [inline dbapi/src/main/java/com/griddynamics/jagger/dbapi/model/rules/TreeViewGroupMetricsToNodeRuleProvider.java]: why SUCCESS_RATE_OLD_ID is still there?

File: dbapi/src/main/java/com/griddynamics/jagger/dbapi/model/rules/TreeViewGroupMetricsToNodeRuleProvider.java [hunk 6]
```diff
@@ -183,6 +180,17 @@ public TreeViewGroupMetricsToNodeRule provide(Map<String, Set<String>> agentName
                 StandardMetricsNamesUtil.THROUGHPUT,
                 regex));
 
+        // Virtual Users
+        regex = "^(" +
+                StandardMetricsNamesUtil.VIRTUAL_USERS_ID + "|" +
+                StandardMetricsNamesUtil.VIRTUAL_USERS + ".*" +
```
Comment (dlatnikov) [inline dbapi/src/main/java/com/griddynamics/jagger/dbapi/model/rules/TreeViewGroupMetricsToNodeRuleProvider.java]: why display name is required in the regexp? why not only id?

File: dbapi/src/main/java/com/griddynamics/jagger/dbapi/parameter/DefaultWorkloadParameters.java [hunk 7]
```diff
@@ -18,7 +18,8 @@
     TIME_LATENCY_PERCENTILE_80("80.0", true, MetricNameDto.Origin.LATENCY_PERCENTILE),
     TIME_LATENCY_PERCENTILE_90("90.0", true, MetricNameDto.Origin.LATENCY_PERCENTILE),
     TIME_LATENCY_PERCENTILE_95("95.0", true, MetricNameDto.Origin.LATENCY_PERCENTILE),
-    TIME_LATENCY_PERCENTILE_99("99.0", true, MetricNameDto.Origin.LATENCY_PERCENTILE);
+    TIME_LATENCY_PERCENTILE_99("99.0", true, MetricNameDto.Origin.LATENCY_PERCENTILE),
+    VIRTUAL_USERS(StandardMetricsNamesUtil.VIRTUAL_USERS, false, MetricNameDto.Origin.VIRTUAL_USERS);
```
Comment (dlatnikov) [inline dbapi/src/main/java/com/griddynamics/jagger/dbapi/parameter/DefaultWorkloadParameters.java]: is this class (DefaultWorkloadParameters) used anywhere in the code?

File: dbapi/src/main/resources/spring/dbapi.config.xml [hunk 8]
```diff
@@ -193,6 +193,19 @@
...
+                    </key>
+                    <array value-type="com.griddynamics.jagger.dbapi.parameter.DefaultWorkloadParameters">
+                        <value>VIRTUAL_USERS</value>
+                    </array>
+                </entry>
+                <entry>
+                    <key>
+                        <bean class="com.griddynamics.jagger.dbapi.parameter.GroupKey">
+                            <constructor-arg index="0">
+                                <util:constant
```
Comment (dlatnikov) [inline dbapi/src/main/resources/spring/dbapi.config.xml]: is this hash map used anywhere in the code?

File: dbapi/src/main/resources/spring/dbapi.config.xml [hunk 9]
```diff
@@ -235,6 +248,15 @@
         </property>
         <property name="origin" value="STANDARD_METRICS"/>
     </bean>
+    <bean id="virtualUsersDto" class="com.griddynamics.jagger.dbapi.dto.MetricNameDto">
```
Comment (dlatnikov) [inline dbapi/src/main/resources/spring/dbapi.config.xml]: is it ever used?

File: dbapi/src/main/resources/spring/dbapi.config.xml [hunk 10]
```diff
@@ -281,22 +303,13 @@
...
-    </bean>
     <util:list id="standardMetricNameDtoList">
         <ref bean="throughputMetricNameDto"/>
         <ref bean="latencyMetricNameDto"/>
         <ref bean="durationMetricNameDto"/>
         <ref bean="successRateMetricNameDto"/>
         <ref bean="samplesMetricNameDto"/>
         <ref bean="latencyStdDevMetricNameDto"/>
-        <ref bean="failsMetricNameDto"/>
+        <ref bean="virtualUsersDto"/>
```
Comment (dlatnikov) [inline dbapi/src/main/resources/spring/dbapi.config.xml]: is it ever used?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/reporting/SummaryReporter.java [hunk 11]
```diff
@@ -168,7 +168,6 @@ private void getData(String sessionId) {
 
             Set<String> standardMetricsIds = new HashSet<String>();
             standardMetricsIds.addAll(StandardMetricsNamesUtil.getAllVariantsOfMetricName(StandardMetricsNamesUtil.THROUGHPUT_ID));
-            standardMetricsIds.addAll(StandardMetricsNamesUtil.getAllVariantsOfMetricName(StandardMetricsNamesUtil.FAIL_COUNT_ID));
             standardMetricsIds.addAll(StandardMetricsNamesUtil.getAllVariantsOfMetricName(StandardMetricsNamesUtil.SUCCESS_RATE_ID));
             standardMetricsIds.addAll(StandardMetricsNamesUtil.getAllVariantsOfMetricName(StandardMetricsNamesUtil.LATENCY_ID));
```
Comment (SokolAndrey) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/reporting/SummaryReporter.java]: What about backward compatibility? If though I can delete all synonyms for all metrics.

File: dbapi/src/main/java/com/griddynamics/jagger/dbapi/model/rules/TreeViewGroupMetricsToNodeRuleProvider.java (hunk 6 above)
Comment (SokolAndrey) [inline dbapi/src/main/java/com/griddynamics/jagger/dbapi/model/rules/TreeViewGroupMetricsToNodeRuleProvider.java]: The same as everywhere

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/SuccessRateAggregatorProvider.java [hunk 12]
```diff
@@ -49,7 +51,7 @@ public void reset() {
 
         @Override
         public String getName() {
-            return "Success rate";
+            return SUCCESS_RATE;
         }
     }
 }
```
Comment (SokolAndrey) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/SuccessRateAggregatorProvider.java]: Done

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/reporting/SummaryReporter.java [hunk 13]
```diff
@@ -168,7 +168,6 @@ private void getData(String sessionId) {
...
             standardMetricsIds.addAll(StandardMetricsNamesUtil.getAllVariantsOfMetricName(StandardMetricsNamesUtil.LATENCY_STD_DEV_ID));
```
Comment (SokolAndrey) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/reporting/SummaryReporter.java]: Added

File: chassis/util/src/main/java/com/griddynamics/jagger/util/StandardMetricsNamesUtil.java [hunk 14]
```diff
@@ -37,24 +44,30 @@
...
+    public static final String SUCCESS_RATE_AGGREGATOR_OK_ID = "Success rate";
+    public static final String SUCCESS_RATE_AGGREGATOR_FAILED_ID = "Number of fails";
 
     //begin: following section is used for docu generation - standard metrics ids
     public static final String THROUGHPUT_ID = "throughput";
     public static final String LATENCY_ID = "avgLatency";
     public static final String LATENCY_STD_DEV_ID = "stdDevLatency";
-    public static final String FAIL_COUNT_ID = "failureCount";
     public static final String SUCCESS_RATE_ID = "successRate";
+    public static final String SUCCESS_RATE_OK_ID = SUCCESS_RATE_ID + "[ " + SUCCESS_RATE_AGGREGATOR_OK_ID + " ]";
```
Comment (dlatnikov) [inline chassis/util/src/main/java/com/griddynamics/jagger/util/StandardMetricsNamesUtil.java]: createAggregatorDisplayNameSuffix from com.griddynamics.jagger.engine.e1.aggregator.workload.MetricLogProcessor

**Pull request merged**